src/Disks/ObjectStorages/MetadataStorageFromDisk.cpp

```diff
@@ -177,12 +179,12 @@ void MetadataStorageFromDiskTransaction::commit()
             toString(state), toString(MetadataFromDiskTransactionState::PREPARING));
 
     {
-        std::lock_guard lock(metadata_storage.metadata_mutex);
+        std::unique_lock lock(metadata_storage.metadata_mutex);
         for (size_t i = 0; i < operations.size(); ++i)
         {
             try
             {
-                operations[i]->execute();
+                operations[i]->execute(lock);
```

Comment:
If I understood correctly you need `metadata_storage.metadata_mutex` to be locked while you're calling `execute()`. Before your changes it was so and now it should be so. Nothing's changed. Did I miss something?

Reply:
Now the whole non atomic operations like hardlink, add blob and set read only became atomic (with one mutex)